### DISCUSSION TITLE
Error running method "end_test" of plugin "tap": I/O operation on closed file.

### DISCUSSION THREAD

**Discussion by marioaag** [avocado-framework/avocado#5749] [Q&A] [answered]
Hi, I'm generating manually the test_suits and the job. In the pass I was just creating one list of suites and running all of them in one shot, that worked without any problem, example:
```
    # Running the test suite
    with Job(config=JOB_CONFIG, test_suites=test_suites) as job:
        status = job.run()
```

Now I'm exploring run using the same job, one test_suit by one test_suit, this with the purpose of do some conditions in the middle  and some other tricks, example:
```
    # Running the test suites
    status = PASS
    job = Job(config=JOB_CONFIG)
    job.setup()
    for test_suite in test_suites:
        job.test_suites = [test_suite]
        if job.run():
            status = FAIL
```

but now I'm getting this error message in the terminal:
```
Error running method "end_test" of plugin "tap": I/O operation on closed file.

Reproduced traceback from: /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/avocado/core/extension_manager.py:224
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/avocado/plugins/tap.py", line 142, in end_test
    self.__write("not ok %s %s", result.tests_run, name)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/avocado/plugins/tap.py", line 77, in __write
    log(msg, *writeargs)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/avocado/plugins/tap.py", line 42, in writeln
    ret = log_file.write(msg + "\n")
ValueError: I/O operation on closed file.
```

I'm using the latest version of avocado (102.0), any idea of how to get rid of this error message? or it should be consider a bug?

The execution by itself looks to be done without problems.

**Comment by marioaag** [comment]
@richtja, do you have any suggestion?

**Reply by richtja** [reply]
Hi @marioaag IMO the main issue is that you are running the same job multiple times with different suites. The Job hasn't been design for such use-case and we don't test is, therefore the results might not be stable. I don't know your use-case, but it is possible for you to run one job for each test_suite? Like this:

```
# Running the test suites
    status = PASS
    
    for test_suite in test_suites:
        job = Job(config=JOB_CONFIG)
        job.setup()
        job.test_suites = [test_suite]
        if job.run():
            status = FAIL
```

**Comment by marioaag** [comment]
IC, thanks for the clarification, I will see how to handle this due running using multiple Jobs will lead to have the results in multiple directories and I was trying to keep all the results in the same directory.

Any suggestion on how to achieve this?

**Reply by richtja** [reply, marked as answer]
I see, unfortunately right now each Job has its own directory. AFAIK there is no easy way how to achieve this.

Since I don't know your exact use-case I am just guessing here, but if you want to run some code between suites, maybe you could use one of these options:

1. You can create your own [SuiteRunner ](https://github.com/avocado-framework/avocado/blob/master/avocado/core/plugin_interfaces.py#L301)which will be based on [avocado SuiteRunner ](https://github.com/avocado-framework/avocado/blob/master/avocado/plugins/runner_nrunner.py#L200)and add your code there. Of course, this is a complex solution and it might be difficult to maintain.
2. We plan to add [suite dependencies](https://github.com/avocado-framework/avocado/issues/5504) in [104](https://github.com/avocado-framework/avocado/milestone/30), which will introduce Pre/Post Suite plugins. Then you could create your own plugin which will run your code between suites. This will be an easier solution, but we don't support it yet. Therefore, you have to wait for implementation of Pre/Post Suite mechanism.
3. You can fix the tap plugin issue and use your initial solution, but since we do not support such behavior, we can't ensure stability of it and you might face other issues in the future.

**Reply by marioaag** [reply]
Thanks so much for all the details, the suite dependencies sounds promising.
